File: concepts/08_async_and_concurrency/02_scheduling_tasks_microtasks_and_futures/nodejs/test_02_ready_queue_boundaries.mjs
// ready queue、完成回调与显式执行边界。
// 共同问题：已经完成的结果何时通知后来观察者；同一队列是否保持登记顺序；
// 创建结果对象是否等同于启动工作。
//
// polyglot-family: async_and_concurrency
// polyglot-concept: scheduling_tasks_microtasks_and_futures
// polyglot-related: languages/nodejs/language/test_019_async_functions_await_microtasks_and_concurrency.mjs

import assert from 'node:assert/strict';
import test from 'node:test';

test('已 fulfilled Promise 的后来 reaction 仍异步排入 job queue', async () => {
  const events = [];
  const result = Promise.resolve(42);

  result.then(() => events.push('reaction'));
  events.push('sync');

  assert.deepEqual(events, ['sync']);
  await Promise.resolve();
  assert.deepEqual(events, ['sync', 'reaction']);
});

test('同一 Promise 的 reactions 按登记顺序执行', async () => {
  const events = [];
  const result = Promise.resolve();

  result.then(() => events.push('first'));
  result.then(() => events.push('second'));
  await result;
  await Promise.resolve();

  assert.deepEqual(events, ['first', 'second']);
});

test('Promise 构造器启动 executor，但 reaction 等待当前栈结束', async () => {
  const events = [];
  const result = new Promise((resolve) => {
    events.push('executor');
    resolve(42);
  });
  result.then(() => events.push('reaction'));
  events.push('sync');

  assert.deepEqual(events, ['executor', 'sync']);
  await result;
  await Promise.resolve();
  assert.deepEqual(events, ['executor', 'sync', 'reaction']);
});

File: concepts/08_async_and_concurrency/02_scheduling_tasks_microtasks_and_futures/rust/test_02_ready_queue_boundaries.rs
// polyglot-family: async_and_concurrency
// polyglot-concept: scheduling_tasks_microtasks_and_futures
// polyglot-related: languages/rust/tests/language/12_async_future_pin_and_cancellation/
// polyglot-related+: test_093_minimal_standard_library_executor.rs
//
// 共同问题：多个已就绪任务的优先级是什么；队列边界能否依赖固定先后。
// 对照观察：std has no ready queue policy；executor chooses poll order，manual polling makes that order explicit.

use std::future::Future;
use std::pin::Pin;
use std::task::{Context, Poll, Waker};

#[test]
fn comparison() {
    let events = std::cell::RefCell::new(Vec::new());
    let mut first = Box::pin(async {
        events.borrow_mut().push("first");
    });
    let mut second = Box::pin(async {
        events.borrow_mut().push("second");
    });
    let mut context = Context::from_waker(Waker::noop());
    assert_eq!(
        Future::poll(Pin::as_mut(&mut second), &mut context),
        Poll::Ready(())
    );
    assert_eq!(
        Future::poll(Pin::as_mut(&mut first), &mut context),
        Poll::Ready(())
    );
    assert_eq!(*events.borrow(), ["second", "first"]);
}
